### DISCUSSION TITLE
Is possible to inject language servers based on treesitter queries?

### DISCUSSION THREAD

**Discussion by olrtg** [neovim/neovim#29477] [Q&A] [answered]
Hey all! I'm the autor of the [emmet-language-server](https://github.com/olrtg/emmet-language-server) and the language server is able to do the following, for example in a `index.html` file:

```html
<body>
  <!-- `ul>li*3` will expand to -->
  <ul>
    <li></li>
    <li></li>
    <li></li>
  </ul>
</body>
```

And in a `index.css` file:

```css
ul {
  /* `w100p` will expand to */
  width: 100%;
}
```

But the problem now is, what if instead my `index.html` is like this instead:

```html
<body>
  <style>
    ul {
      /* `w100p` will not expand because the language server doesn't know that now it has to change to "css" mode an give completions for css rule abbreviations */
    }
  </style>

  <!-- `ul>li*3` will correctly expand because we are indeed in a html file -->
  <ul>
    <li></li>
    <li></li>
    <li></li>
  </ul>
</body>
```

And I already have the [nvim-emmet](https://github.com/olrtg/nvim-emmet/) which offers a bit more integration with the language server, but I was wondering, is there a way to notify from the client to the `emmet_language_server` depending if I'm in a `style` tag or not that it needs to change its config? I was thinking on leveraging treesitter to do that but I don't know if that's even possible with our current API. I saw that there are a couple of methods (`vim.lsp.buf_request_all()` and `vim.lsp.buf_request_sync()`) but those methods make requests for all language servers and I just want to be able to talk to the `emmet_language_server`.

Thanks in advance! and please ask more questions if you need more context.

**Comment by clason** [comment]
This is a shortcoming of the protocol itself: parsers work with full buffers (in Neovim parlance) and there is no way to tell them to only look at parts of them. It is up to the server to ignore parts it can't or doesn't need to handle.

In short: **No**, it's not possible at all, and nothing can be done without a significant change to the specification itself.

**Reply by HyperAfnan** [reply]
isn't there something that can help language server to integrate with treesitter?

**Reply by clason** [reply]
No.

**Reply by olrtg** [reply]
@clason thanks a lot for the answer! If I'm not mistaken the protocol allows you to change the server configuration on the fly, but I'm still learning about this topic so I don't know if I'm correct.

With that said, don't you think it can be possible to detect when you are in certain parts of the buffer to notify the server that it should change it's config? With a plugin or some mechanism built into neovim itself?

**Reply by clason** [reply]
That is completely irrelevant to your question, though, which the protocol just doesn't apply to (yet).

**Reply by olrtg** [reply]
Yeah! sorry, it was unrelated. I'm just trying to see if there's anything within my reach to solve this specific problem. But just to make sure if I understood correctly and I'm making the right assumptions:

1. There's currently no way to talk only to a specific language server. So you can't really send notifications to a single server.
2. It depends on the server to ignore parts that don't need or can't handle. This comes with a question again I guess, can a language server if it sees a `<style>` tag for example provide other types of completions? But I guess I'll probably have to parse the file in some way.
3. You cannot watch with a plugin or something like that a certain part of the buffer with a Tree-sitter query and notify the server that it should make the switch. I guess you can discard this if point 1 is true.

Again, thank you so much for the answers and for your time!

**Reply by clason** [reply, marked as answer]
1. is false (and unrelated)
2. is correct, and anything beyond that depends on the server and cannot be answered in general
3. is unnecessary since the server always gets full buffer updates, so it has all the information it needs.

The problem here is that the protocol does not cover "partial files", so servers can't do this (in a standard way). Until that is changed, there's nothing you can do.
